File: lang/ruby/src/fun/vac/algorithms/number_theory/PrimeSieves.rb
module PrimeSieves

    def self.sieveOfEratosthenes(n)
        if n < 2
            return Array.new(0)
        end

        size = (n + 1) >> 1

        arr = Array.new(size, false)

        numPrimes = size
        i, bound = 3, Math.sqrt(n).to_i() + 1
        while i < bound
            if arr[i >> 1]
                i += 2
                next
            end
            j = i * i
            while j <= n
                unless arr[j >> 1]
                    arr[j >> 1] = true
                    numPrimes -= 1
                end
                j += i << 1
            end
            i += 2
        end

        primes = Array.new(numPrimes)

        primes[0] = 2

        curr = 1
        i, bound = 3, n + 1
        while i < bound
            unless arr[i >> 1]
                primes[curr] = i
                curr += 1
            end
            i += 2
        end

        primes
    end
end

File: go/muse/algorithms/number_theory/prime_sieves/prime_sieves.go
package prime_sieves

import "math"

func SieveOfEratosthenes(n int) []int {
	if n < 2 {
		return make([]int, 0)
	}

	size := (n + 1) >> 1

	arr := make([]bool, size)

	numPrimes := size
	for i, bound := 3, int(math.Sqrt(float64(n)))+1; i < bound; i += 2 {
		if arr[i>>1] {
			continue
		}
		for j := i * i; j <= n; j += i << 1 {
			if !arr[j>>1] {
				arr[j>>1] = true
				numPrimes--
			}
		}
	}

	primes := make([]int, numPrimes)

	primes[0] = 2

	curr := 1
	for i, bound := 3, n+1; i < bound; i += 2 {
		if !arr[i>>1] {
			primes[curr] = i
			curr++
		}
	}

	return primes
}
